checkers/sortSlice_checker.go

```diff
@@ -0,0 +1,135 @@
+package checkers
+
+import (
+	"go/ast"
+	"go/token"
+
+	"github.com/go-critic/go-critic/checkers/internal/lintutil"
+	"github.com/go-lintpack/lintpack"
+	"github.com/go-lintpack/lintpack/astwalk"
+	"github.com/go-toolsmith/astcast"
+	"github.com/go-toolsmith/astequal"
+	"github.com/go-toolsmith/typep"
+	"golang.org/x/tools/go/ast/astutil"
+)
+
+func init() {
+	var info lintpack.CheckerInfo
+	info.Name = "sortSlice"
+	info.Tags = []string{"diagnostic", "experimental"}
+	info.Summary = "Detects suspicious sort.Slice calls"
+	info.Before = `sort.Slice(xs, func(i, j) bool { return keys[i] < keys[j] })`
+	info.After = `sort.Slice(kv, func(i, j) bool { return kv[i].key < kv[j].key })`
+
+	collection.AddChecker(&info, func(ctx *lintpack.CheckerContext) lintpack.FileWalker {
+		return astwalk.WalkerForExpr(&sortSliceChecker{ctx: ctx})
+	})
+}
+
+type sortSliceChecker struct {
+	astwalk.WalkHandler
+	ctx *lintpack.CheckerContext
+}
+
+func (c *sortSliceChecker) VisitExpr(expr ast.Expr) {
+	call := astcast.ToCallExpr(expr)
+	if len(call.Args) != 2 {
+		return
+	}
+	switch qualifiedName(call.Fun) {
+	case "sort.Slice", "sort.SliceStable":
+		// OK.
+	default:
+		return
+	}
+
+	slice := c.unwrapSlice(call.Args[0])
+	lessFunc, ok := call.Args[1].(*ast.FuncLit)
+	if !ok {
+		return
+	}
+	if !typep.SideEffectFree(c.ctx.TypesInfo, slice) {
+		return // Don't check unpredictable slice values
+	}
+
+	ivar, jvar := c.paramIdents(lessFunc.Type)
+	if ivar == nil || jvar == nil {
+		return
+	}
+
+	if len(lessFunc.Body.List) != 1 {
+		return
+	}
+	ret, ok := lessFunc.Body.List[0].(*ast.ReturnStmt)
+	if !ok {
+		return
+	}
+	cmp := astcast.ToBinaryExpr(astutil.Unparen(ret.Results[0]))
+	if !typep.SideEffectFree(c.ctx.TypesInfo, cmp) {
+		return
+	}
+	switch cmp.Op {
+	case token.LSS, token.LEQ, token.GTR, token.GEQ:
+		// Both cmp.X and cmp.Y are expected to be some expressions
+		// over the `slice` expression. In the simplest case,
+		// it's a `slice[i] <op> slice[j]`.
+		if !c.containsSlice(cmp.X, slice) && !c.containsSlice(cmp.Y, slice) {
+			c.warnSlice(cmp, slice)
+		}
+
+		// This one is more about the style, but can reveal potential issue
+		// or misprint in sorting condition.
+		// We give a warn if X contains indexing with `i` index and Y
+		// contains indexing with `j`.
+		if c.containsIndex(cmp.X, jvar) && c.containsIndex(cmp.Y, ivar) {
+			c.warnIndex(cmp, ivar, jvar)
+		}
+	}
+}
+
+func (c *sortSliceChecker) paramIdents(e *ast.FuncType) (*ast.Ident, *ast.Ident) {
```

Comment:
?
```
fields := e.Params.List
if len(fields) == 2 {
    return fields.Names[0], fields.Names[1]
}
return nil, nil
```

Reply:
This won't handle the case where both idents are a single field.
```
i, j int - 1 field with 2 names
i int, j int - 2 fields, each with 1 name
```

Reply:
oh! got it, nice.